### PR TITLE
feat: Performance monitoring

### PR DESCRIPTION
**Pull request by bruno-garcia**
This PR is not going to receive commits directly. It'll serve as `base` for PRs related to performance monitoring and will be merged into `master` **without squash**.

### CODE REVIEW

**Comment by marandaneto** [comment]
let's also pull master into `feat/performance-monitoring` sometimes so we resolve smaller junks of conflicts if any

**Comment by philipphofmann** [comment]
I configured Github Actions with https://github.com/getsentry/sentry-cocoa/pull/909/commits/e08c892d1f526b3c134e08e1074677926916f986 for this branch.

**Comment by bruno-garcia** [comment]
> let's also pull master into `feat/performance-monitoring` sometimes so we resolve smaller junks of conflicts if any

Yes but this time maybe rebase, so we dont' end up with a lot of merge commits into main when we merge there without squash

**Review by philipphofmann — CHANGES_REQUESTED** [review]
I know this is WIP, and we want to see something happening, but before merging this PR, let's add tests for the new functionality. Codecov tells us the coverage is going down.

File: Sources/Sentry/SentryTransaction.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTransaction.m:28]: `m`: Why do we set the timestamp here? Shouldn't this only happen in `finish`?

File: Sources/Sentry/Public/SentryEvent.h
Comment (bruno-garcia) [reply Sources/Sentry/Public/SentryEvent.h:27]: We only know the 'end time' of a transaction when it ends.

File: Sources/Sentry/SentryTransaction.m
Comment (bruno-garcia) [reply Sources/Sentry/SentryTransaction.m:28]: Good catch. This should definitely be fixed.
@marandaneto is this similar to what we discussed about not compensating when this is `nil`? 
Meaning we send `nil` to the server if it wasn't `finished` properly?

@brustolin Or does the server complain and this was a workaround or something?

Comment (brustolin) [reply Sources/Sentry/SentryTransaction.m:28]: I copied this behavior from SentryEvent serialization before I start to cascade serialization. I do think this does not belong here, but SentryEvent serialization will do it anyway.

Comment (brustolin) [reply Sources/Sentry/SentryTransaction.m:28]: This was copied over from event, before transaction inherited from event. Now we don't need it anymore. But SentryEvent will do it anyway, so the timestamp is not only set at finish()

Comment (bruno-garcia) [reply Sources/Sentry/SentryTransaction.m:28]: Sounds like we should address this. this value should be set in the SentryEvent constructor (for errors) but not for Transactions.

Comment (marandaneto) [reply Sources/Sentry/SentryTransaction.m:28]: yeah we can't do that, as @bruno-garcia said.

a SentryEvent sets a `timestamp` on ctor, when the instance is created (always, consider non-nullable).

a Transaction has a nullable `timestamp` during creation.
a Transaction sets a `timestamp` when `finish` is called (always, consider non-nullable).
ps: if the user forgot to `transaction.finish()`, we should not set a `timestamp` automatically, it should send as it is, same for Spans

File: Sources/Sentry/SentryTransaction.m
Comment (philipphofmann) [inline Sources/Sentry/SentryTransaction.m:18]: `m`: @marandaneto shouldn't this be a TransactionContext? I know it is different in Java.

Comment (marandaneto) [reply Sources/Sentry/SentryTransaction.m:18]: TransactionContext inherits from SentrySpanContext, the 2 fields, in addition, are used prior to this class, so at runtime, the transaction only needs to know the `SentrySpanContext` fields.
it could be, but there's no reason either.

**Comment by philipphofmann** [comment]
> This PR is not going to receive commits directly. It'll serve as `base` for PRs related to performance monitoring and will be merged into `master` **without squash**.

What about squashing this whole PR as just one commit `Performance API` to master? So we can just merge like crazy to this PR and don't have to rebase all the time? @bruno-garcia @marandaneto and @brustolin

**Comment by marandaneto** [comment]
> > This PR is not going to receive commits directly. It'll serve as `base` for PRs related to performance monitoring and will be merged into `master` **without squash**.
> 
> What about squashing this whole PR as just one commit `Performance API` to master? So we can just merge like crazy to this PR and don't have to rebase all the time? @bruno-garcia @marandaneto and @brustolin

I prefer not squashing, history might be valuable when investigating things

**Comment by philipphofmann** [comment]
What about merging this to master, as we have 7.0.0-alpha.0 already? @bruno-garcia, @marandaneto

**Comment by marandaneto** [comment]
> What about merging this to master, as we have 7.0.0-alpha.0 already? @bruno-garcia, @marandaneto

there are still a few moving bits, I'd wait for the sampling context PR to be merged and work a bit on samples.
right now the samples create a transaction, start an async dispatch, and finish the transaction.

I'd prefer to have a sample that creates a transaction, start a child, finish as errored, capture an event in-between, see if it associate its data, etc, right now there are only unit tests but no real usage/similar to the reality to be sure that everything renders nicely in the UI, other than that, I'd be totally in for making an alpha, what do you think @bruno-garcia @philipphofmann @brustolin ?

**Comment by bruno-garcia** [comment]
> > What about merging this to master, as we have 7.0.0-alpha.0 already? @bruno-garcia, @marandaneto
> 
> there are still a few moving bits, I'd wait for the sampling context PR to be merged and work a bit on samples.
> right now the samples create a transaction, start an async dispatch, and finish the transaction.
> 
> I'd prefer to have a sample that creates a transaction, start a child, finish as errored, capture an event in-between, see if it associate its data, etc, right now there are only unit tests but no real usage/similar to the reality to be sure that everything renders nicely in the UI, other than that, I'd be totally in for making an alpha, what do you think @bruno-garcia @philipphofmann @brustolin ?

Makes total sense to me! Sample would be great.

**Comment by marandaneto** [comment]
@brustolin please resolve the open comments if they are already resolved, so it's easier to review it again, thanks

**Comment by marandaneto** [comment]
changelog requires merging

**Comment by marandaneto** [comment]
@bruno-garcia can we check if the sampling decision follows the correct format based on the discussion yesterday? I quite didn't fully get it yet and it'd be nice to double-check this before releasing, even if it's wrong, it'd be an easy fix.

File: Sources/Sentry/Public/SentryEvent.h
Comment (philipphofmann) [reply Sources/Sentry/Public/SentryEvent.h:34]: Okay I added it as a breaking change to the changelog.

**Comment by philipphofmann** [comment]
> @bruno-garcia can we check if the sampling decision follows the correct format based on the discussion yesterday? I quite didn't fully get it yet and it'd be nice to double-check this before releasing, even if it's wrong, it'd be an easy fix.

@marandaneto, I created an issue for this https://github.com/getsentry/sentry-cocoa/issues/986. I think it's fine to release the alpha without this fix.

**Review by philipphofmann — APPROVED** [review]
For an alpha LGTM. I don't want to keep this on an extra branch any longer. Is there anything stopping us from merging this @bruno-garcia, @marandaneto, and @brustolin?

**Comment by marandaneto** [comment]
> > @bruno-garcia can we check if the sampling decision follows the correct format based on the discussion yesterday? I quite didn't fully get it yet and it'd be nice to double-check this before releasing, even if it's wrong, it'd be an easy fix.
> 
> @marandaneto, I created an issue for this #986. I think it's fine to release the alpha without this fix.

it's not about that, it's something else.

**Review by marandaneto — APPROVED** [review]
LGTM

File: .github/workflows/buildandtest.yml
Comment (marandaneto) [inline .github/workflows/buildandtest.yml:7]: @brustolin maybe this could be removed now?

File: CHANGELOG.md
Comment (marandaneto) [inline CHANGELOG.md:11]: maybe a single entry here would make more sense, but all the PRs ids between ()

**Pull request merged**
